### PR TITLE
Add HttpWebRequest based transport

### LINKED ISSUE #13477: Add support for sync HttpWebRequest-based transport on .NET Framework

### PR DESCRIPTION
**Pull request by pakrym**
Adds a HttpWebRequest-based transport that's internal and enabled by default on .NET Framework.

Adds and environment variable `AZURE_CORE_DISABLE_HTTPWEBREQUESTTRANSPORT` and AppContext switch `Azure.Core.Pipeline.DisableHttpWebRequestTransport` to be able to disable the new transport.

Fixes a few bugs with timeouts in sync methods.

Converts HttpClientTransport-test to a more functional style that run against all transport implementations and includes sync runs.

Fixes: https://github.com/Azure/azure-sdk-for-net/issues/13477

### CODE REVIEW

**Review by christothes — APPROVED** [review]

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 1]
```diff
@@ -133,6 +133,10 @@ jobs:
         Windows_NetFramework:
           OSVmImage: "windows-2019"
           TestTargetFramework: net461
+        Windows_NetFramework_ProjectReferences:
```
Comment (pakrym) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:136]: @weshaggard FYI.

We are adding .NET Framework specific code that's on a very hot path adding a configuration to make sure it's tested before the release.

File: sdk/core/Azure.Core/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,15 @@
 
 ## 1.4.0-preview.1 (Unreleased)
 
+### Added
+- HttpWebRequest-based transport implementation. Enabled by-default on .NET Framework. Can be disabled using `AZURE_CORE_DISABLE_HTTPWEBREQUESTTRANSPORT` environment variable or `Azure.Core.Pipeline.DisableHttpWebRequestTransport` AppContext switch. To use the app context switch add the following snippet to your `.csproj`:
```
Comment (heaths) [inline sdk/core/Azure.Core/CHANGELOG.md:6]: Is it worth mentioning what would be used if disabled?

File: sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+        private async ValueTask ProcessInternal(HttpMessage message, bool async)
+        {
+            var request = CreateRequest(message.Request);
+            using var registration = message.CancellationToken.Register(state => ((HttpWebRequest) state).Abort(), request);
+            try
+            {
+                if (message.Request.Content != null)
+                {
+                    if (request.ContentLength == -1 &&
```
Comment (heaths) [inline sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs:56]: Would it be safer to just check `< 0`?

File: sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        }
+
+        private HttpWebRequest CreateRequest(Request messageRequest)
+        {
+            var request = WebRequest.CreateHttp(messageRequest.Uri.ToUri());
+            request.Method = messageRequest.Method.Method;
+            request.Proxy = _proxy;
+            foreach (var messageRequestHeader in messageRequest.Headers)
+            {
+                if (messageRequestHeader.Name == HttpHeader.Names.ContentLength)
```
Comment (heaths) [inline sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs:96]: HTTP headers are case-insensitive. May want to do a `OrdinalIgnoreCase` comparison on these (below as well).

**Review by heaths — APPROVED** [review]

File: sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs (hunk 4 above)
Comment (pakrym) [reply sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs:96]: Great point.

**Review by heaths — APPROVED** [review]

File: sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs (hunk 3 above)
Comment (pakrym) [reply sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs:56]: I'll keep it -1 to know if there are any other magic numbers :)

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml (hunk 1 above)
Comment (weshaggard) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:136]: Do you intend for this to stay longer term? If it doesn't add too much cost I'm fine with it staying but I know we are going to start testing a larger matrix soon and will likely look for candidates to trim on default builds.

Comment (pakrym) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:136]: We would be able to remove it after new features ship as a package.

**Review by tg-msft — APPROVED** [review]

File: sdk/core/Azure.Core/CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,17 @@
...
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/core/Azure.Core/CHANGELOG.md:6]: I don't know a good reason yet just preparing to give people a workaround if they need too.

File: sdk/core/Azure.Core/src/ClientOptions.cs [hunk 6]
```diff
@@ -13,7 +13,7 @@ namespace Azure.Core
     /// </summary>
     public abstract class ClientOptions
     {
-        private HttpPipelineTransport _transport = HttpClientTransport.Shared;
+        private HttpPipelineTransport _transport = GetDefaultTransport();
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/core/Azure.Core/src/ClientOptions.cs:16]: SetSwitch can be called at any part of the application so I think it's better to re-fetch.

File: sdk/core/Azure.Core/src/ClientOptions.cs [hunk 7]
```diff
@@ -80,5 +80,33 @@ public void AddPolicy(HttpPipelinePolicy policy, HttpPipelinePosition position)
...
+                    {
+                        ret = string.Equals("true", switchValue, StringComparison.InvariantCultureIgnoreCase) ||
+                              switchValue.Equals("1", StringComparison.InvariantCultureIgnoreCase);
+                    }
+                }
+
+                return ret;
+            }
+
+            if (!GetSwitchValue("Azure.Core.Pipeline.DisableHttpWebRequestTransport", "AZURE_CORE_DISABLE_HTTPWEBREQUESTTRANSPORT"))
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/core/Azure.Core/src/ClientOptions.cs:103]: We can but that requires changing all client constructor calls which is not very practical.

File: sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+    /// </summary>
+    internal class HttpWebRequestTransport : HttpPipelineTransport
+    {
+        public static readonly HttpWebRequestTransport Shared = new HttpWebRequestTransport();
+        private readonly IWebProxy? _proxy;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="HttpWebRequestTransport"/>
+        /// </summary>
+        public HttpWebRequestTransport()
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs:28]: You can somewhat use the ServicePointManager to configure things but it's pretty hard to use to target a specific client. In general I'd like to start small with this feature and keep it internal until we gain more confidence.

[the remark this answers is not in the document]
Comment (tg-msft) [reply sdk/core/Azure.Core/src/Pipeline/HttpWebRequestTransport.cs:28]: I was thinking more like setting client certificates than validating server certificates.  Maybe we could take a `Func<Uri, HttpWebRequest>` as an optional parameter defaulting to `WebRequest.CreateHttp` at some point down the road.

**Comment by KrzysztofCwalina** [comment]
cc: @stephentoub

**Pull request merged**